File: tools/github/npm-bootstrap-preflight.mjs
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import {
  assertPackageIsUnpublished,
  validateBootstrapManifest,
  validateBootstrapRequest,
  validateDownloadedArtifact,
  validateReleaseCommit,
  verifyReleaseRunEvidence,
} from "./npm-bootstrap-contract.mjs";

function runGit(args, acceptedExitCodes = [0]) {
  const result = spawnSync("git", args, {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    timeout: 30_000,
  });
  if (result.error?.code === "ETIMEDOUT") {
    throw new Error(`git ${args.join(" ")} timed out after 30000ms`);
  }
  if (result.error != null) throw result.error;
  if (result.signal != null) {
    throw new Error(`git ${args.join(" ")} was terminated by ${result.signal}`);
  }
  if (!acceptedExitCodes.includes(result.status)) {
    const detail = [result.stdout, result.stderr].filter(Boolean).join("\n").trim();
    throw new Error(
      [`git ${args.join(" ")} failed with exit ${result.status}`, detail]
        .filter(Boolean)
        .join("\n"),
    );
  }
  return result;
}

function readTaggedFile(tagSha, filePath) {
  return runGit(["show", `${tagSha}:${filePath}`]).stdout;
}

/** @param {Record<string, string>} outputs */
function writeOutputs(outputs) {
  const outputPath = process.env.GITHUB_OUTPUT;
  if (!outputPath) throw new Error("GITHUB_OUTPUT is required for npm bootstrap preflight");
  fs.appendFileSync(
    outputPath,
    Object.entries(outputs)
      .map(([name, value]) => `${name}=${value}\n`)
      .join(""),
  );
}

async function runPreflight(env = process.env) {
  const request = validateBootstrapRequest({
    tagName: env.RELEASE_TAG_NAME ?? "",
    packagePath: env.BOOTSTRAP_PACKAGE_PATH ?? "",
    releaseRunId: env.RELEASE_RUN_ID ?? "",
    workflowRef: env.GITHUB_REF ?? "",
    workflowSha: env.GITHUB_SHA ?? "",
  });

  runGit(["fetch", "--quiet", "--no-tags", "origin", "+refs/heads/main:refs/remotes/origin/main"]);
  runGit([
    "fetch",
    "--quiet",
    "--no-tags",
    "origin",
    `+refs/tags/${request.tagName}:refs/tags/${request.tagName}`,
  ]);
  const tagSha = runGit(["rev-parse", `refs/tags/${request.tagName}^{commit}`]).stdout.trim();
  const mainSha = runGit(["rev-parse", "refs/remotes/origin/main"]).stdout.trim();
  const mainFirstParentHistory = runGit(["rev-list", "--first-parent", "refs/remotes/origin/main"])
    .stdout.trim()
    .split(/\r?\n/);
  validateReleaseCommit({
    tagSha,
    workflowSha: request.workflowSha,
    mainSha,
    isOnFirstParent: mainFirstParentHistory.includes(tagSha),
  });

  const packageManifest = readTaggedFile(tagSha, `${request.packagePath}/package.json`);
  const version = validateBootstrapManifest({
    ...request,
    tagSha,
    cargoToml: readTaggedFile(tagSha, "Cargo.toml"),
    packageManifest,
  });
  await verifyReleaseRunEvidence({
    apiUrl: env.GITHUB_API_URL ?? "https://api.github.com",
    repository: env.GITHUB_REPOSITORY ?? "",
    token: env.GITHUB_TOKEN ?? "",
    releaseRunId: request.releaseRunId,
    tagName: request.tagName,
    tagSha,
    artifactName: request.artifactName,
  });
  await assertPackageIsUnpublished(request.packageName);
  writeOutputs({
    artifact_name: request.artifactName,
    package_name: request.packageName,
    package_path: request.packagePath,
    release_run_id: request.releaseRunId,
    tag_sha: tagSha,
    version,
  });
  console.log(
    `npm bootstrap preflight passed for ${request.packageName}@${version} from ${request.tagName} (${tagSha}).`,
  );
}

async function runRegistryRecheck(env = process.env) {
  const request = validateBootstrapRequest({
    tagName: env.RELEASE_TAG_NAME ?? "",
    packagePath: env.BOOTSTRAP_PACKAGE_PATH ?? "",
    releaseRunId: env.RELEASE_RUN_ID ?? "",
    workflowRef: env.GITHUB_REF ?? "",
    workflowSha: env.GITHUB_SHA ?? "",
  });
  await assertPackageIsUnpublished(request.packageName);
  console.log(`${request.packageName} is still absent from npm immediately before first publish.`);
}

function runArtifactValidation(env = process.env) {
  const artifactPath = env.BOOTSTRAP_ARTIFACT_PATH ?? "";
  if (artifactPath !== "bootstrap-package") {
    throw new Error(
      `Bootstrap artifact path must be bootstrap-package, got ${artifactPath || "(empty)"}`,
    );
  }
  const expectedName = env.EXPECTED_PACKAGE_NAME ?? "";
  const expectedVersion = env.EXPECTED_PACKAGE_VERSION ?? "";
  validateDownloadedArtifact({
    packageManifest: fs.readFileSync(path.join(artifactPath, "package.json"), "utf8"),
    expectedName,
    expectedVersion,
  });
  console.log(`Validated downloaded Release artifact ${expectedName}@${expectedVersion}.`);
}

const entrypoint = process.argv[1]
  ? fileURLToPath(import.meta.url) === path.resolve(process.argv[1])
  : false;
if (entrypoint) {
  const [command = "preflight", extra] = process.argv.slice(2);
  Promise.resolve()
    .then(() => {
      if (extra != null || !["artifact", "preflight", "registry-recheck"].includes(command)) {
        throw new Error(
          "Usage: rust-script tools/commands/ci/github/npm-bootstrap-preflight.rs [artifact|preflight|registry-recheck]",
        );
      }
      if (command === "preflight") return runPreflight();
      return command === "artifact" ? runArtifactValidation() : runRegistryRecheck();
    })
    .catch((error) => {
      console.error(error instanceof Error ? error.message : String(error));
      process.exit(1);
    });
}

File: tools/commands/ci/github/npm-bootstrap-preflight.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! base64 = "0.22"
//! regex = "1"
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! sha2 = "0.10"
//! tempfile = "3"
//!
//! [package]
//! edition = "2024"
//! ```

#[path = "../../../support/common.rs"]
mod common;
#[path = "../../../support/release/npm_bootstrap.rs"]
mod npm_bootstrap;

use std::{env, process::ExitCode};

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let args = env::args().skip(1).collect::<Vec<_>>();
    let command = match args.as_slice() {
        [] => "preflight",
        [command] if matches!(command.as_str(), "artifact" | "preflight" | "registry-recheck") => {
            command
        }
        [command, rest @ ..] if command == "__contract" => return contract(rest),
        _ => {
            return Err(
                "Usage: rust-script tools/commands/ci/github/npm-bootstrap-preflight.rs [artifact|preflight|registry-recheck]"
                    .to_string(),
            )
        }
    };
    let env = npm_bootstrap::env_map();
    match command {
        "preflight" => npm_bootstrap::run_preflight(&env),
        "artifact" => npm_bootstrap::run_artifact_validation(&env),
        "registry-recheck" => npm_bootstrap::run_registry_recheck(&env),
        _ => unreachable!(),
    }
}

fn contract(args: &[String]) -> Result<(), String> {
    let [mode, payload] = args else {
        return Err("__contract requires <mode> <json>".to_string());
    };
    let value: serde_json::Value =
        serde_json::from_str(payload).map_err(|error| format!("invalid contract JSON: {error}"))?;
    match mode.as_str() {
        "request" => {
            let request = npm_bootstrap::validate_bootstrap_request(
                value
                    .get("tagName")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("packagePath")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("releaseRunId")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("workflowRef")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("workflowSha")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
            )?;
            println!(
                "{}",
                serde_json::to_string(&request).map_err(|error| error.to_string())?
            );
        }
        "manifest" => {
            let version = npm_bootstrap::validate_bootstrap_manifest(
                value
                    .get("tagName")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("tagSha")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("packagePath")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("packageName")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("cargoToml")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("packageManifest")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
            )?;
            println!("{version}");
        }
        "release-commit" => npm_bootstrap::validate_release_commit(
            value
                .get("tagSha")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("workflowSha")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("mainSha")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("isOnFirstParent")
                .and_then(serde_json::Value::as_bool)
                .unwrap_or(false),
        )?,
        "downloaded-artifact" => npm_bootstrap::validate_downloaded_artifact(
            value
                .get("packageManifest")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("expectedName")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("expectedVersion")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
        )?,
        "release-run" => npm_bootstrap::validate_release_run(
            value.get("run").unwrap_or(&serde_json::Value::Null),
            value
                .get("releaseRunId")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("repository")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("tagName")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("tagSha")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
        )?,
        "release-jobs" => {
            let jobs = value
                .get("jobs")
                .and_then(serde_json::Value::as_array)
                .ok_or_else(|| "jobs must be an array".to_string())?;
            npm_bootstrap::validate_release_jobs(jobs)?;
        }
        "release-artifact" => {
            let artifacts = value
                .get("artifacts")
                .and_then(serde_json::Value::as_array)
                .ok_or_else(|| "artifacts must be an array".to_string())?;
            npm_bootstrap::validate_release_artifact(
                artifacts,
                value
                    .get("artifactName")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("releaseRunId")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("tagName")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                value
                    .get("tagSha")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
            )?;
        }
        "registry-response" => npm_bootstrap::validate_registry_response(
            value
                .get("packageName")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            value
                .get("status")
                .and_then(serde_json::Value::as_u64)
                .ok_or_else(|| "status is required".to_string())? as u16,
        )?,
        mode => return Err(format!("unknown contract mode: {mode}")),
    }
    Ok(())
}
